### PR TITLE
[API] Fix requests to Projects Leader fail with `Connection Aborted`

### PR DESCRIPTION
**Pull request by quaark**
Moved the "request with retries" mechanism to a general MLRun util, so it can be used in multiple places.
Now in addition to using it in the SDK for requests against the API:
- The API now uses it for requests against the project leader.
- The opa client uses it for requests against opa.
- The nuclio client uses it for requests against nuclio.
- The worker uses it for requests against the chief.

### CODE REVIEW

File: mlrun/api/utils/clients/iguazio.py [hunk 1]
```diff
@@ -60,13 +59,10 @@ class Client(
 ):
     def __init__(self) -> None:
         super().__init__()
-        http_adapter = requests.adapters.HTTPAdapter(
-            max_retries=urllib3.util.retry.Retry(total=3, backoff_factor=1),
-            pool_maxsize=int(mlrun.mlconf.httpdb.max_workers),
+        self._session = mlrun.utils.HTTPSessionWithRetry(
+            retry_on_exception=mlrun.mlconf.httpdb.projects.retry_leader_request_on_exception
+            == "enabled"
```
Comment (Tankilevitch) [inline mlrun/api/utils/clients/iguazio.py:64]: lets create a class for the possible flags, search for `....FeatureFlag` classes for inference.

File: mlrun/utils/http.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    def __init__(
+        self,
+        max_retries=DEFAULT_RETRY_COUNT,
+        retry_backoff_factor=DEFAULT_RETRY_BACKOFF,
+        retry_on_exception=True,
+        retry_on_status=True,
+        retry_on_post=False,
+    ):
```
Comment (Tankilevitch) [inline mlrun/utils/http.py:33]: Please add docstring for the class

File: mlrun/utils/http.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        if retry_on_status:
+            http_adapter = requests.adapters.HTTPAdapter(
+                max_retries=urllib3.util.retry.Retry(
+                    total=self.max_retries,
+                    backoff_factor=self.retry_backoff_factor,
+                    status_forcelist=[500, 502, 503, 504],
+                    method_whitelist=False
+                    if retry_on_post
+                    else urllib3.util.retry.Retry.DEFAULT_METHOD_WHITELIST,
```
Comment (Tankilevitch) [inline mlrun/utils/http.py:48]: Maybe should move to helper func or add comments, choose whatever you think is clearer.

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, this change affects all our outbounding requests, please add coverage for all affected clients.

File: mlrun/utils/http.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                if not self.retry_on_exception:
+                    logger.warning(
+                        f"{method} {url} request failed, http retries disabled, raising exception: {exc}",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
+                    raise exc
```
Comment (theSaarco) [inline mlrun/utils/http.py:87]: This log block is repeated 2.5 times - twice here and 3rd time with less parameters if it's non-retryable (but could have the same set of parameters there as well). Should make this an internal function or just extract it to a function, rather than repeating the same code again and again.

File: mlrun/utils/http.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                if self.verbose:
+                    logger.debug(
+                        f"{method} {url} request failed on retryable exception, "
+                        f"retrying in {self.retry_backoff_factor} seconds",
+                        exception_type=type(exc),
+                        exception_message=str(exc),
+                        retry_interval=self.retry_backoff_factor,
+                        retry_count=retry_count,
+                        max_retries=self.max_retries,
+                    )
```
Comment (theSaarco) [inline mlrun/utils/http.py:125]: And a fourth time here (only now it's `debug` rather than `warning` - could also use the same internal function, as long as it has a parameter saying whether to use debug or warning...

File: mlrun/utils/http.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+import requests
+import requests.adapters
+import urllib3.util.retry
+
+from ..config import config
+from . import logger
+
+DEFAULT_RETRY_COUNT = 3
+DEFAULT_RETRY_BACKOFF = 1
```
Comment (theSaarco) [inline mlrun/utils/http.py:11]: Now that it's a centralized config and used in several places, I think it makes sense to put these parameters in the general config. This will enable fine-tuning in the field if needed.

Also - why did you rename `HTTP_RETRY_BACKOFF_FACTOR` back to `HTTP_RETRY_BACKOFF`?

**Review by theSaarco — APPROVED** [review]
Great idea, and good execution. I have some comments which are mostly style-related (repeated code etc.).
